### PR TITLE
WIP real-valued functional

### PR DESCRIPTION
**Pull request by miallo**
I started working on the implementation of the range of Functional as mentioned in #1328.

If I am not mistaken the old implementation of FunctionalRightVectorMult leads to the functional “forgetting” its  grad_lipschitz and linear properties.

I tagged some places in functional.py with “# HELP” where I am especially unsure about what I am doing.

I did NOT implement any kind of typecasting anywhere yet but just tried to make sure that the range property gets passed along.

### CODE REVIEW

File: odl/solvers/functional/functional.py [hunk 1]
```diff
@@ -686,7 +693,12 @@ def __init__(self, func, vector):
                             ''.format(func))
 
         OperatorRightVectorMult.__init__(self, operator=func, vector=vector)
-        Functional.__init__(self, space=func.domain)
+        # HELP: before it was Functional.__init__(self, domain=func.domain).
+        # But doesn't this loose the information on grad_lipschitz and linear?
```
Comment (aringh) [inline odl/solvers/functional/functional.py]: Yes, it seems like the previous implementation forgets the linearity and lipschitz constant for the gradient. However, wouldn't the latter change for this functional? Anyway, we are thinking of removing `grad_lipschitz` since it is not currently used anywhere. See #1300

File: odl/solvers/functional/functional.py [hunk 2]
```diff
@@ -1020,11 +1033,11 @@ def __init__(self, func, quadratic_coeff=0, linear_term=None,
             raise ValueError(
                 "Complex-valued `constant` coefficient is not supported.")
         self.__constant = constant.real
-
+        #HELP: Need to check for range? The + should take care of it, right?
```
Comment (aringh) [inline odl/solvers/functional/functional.py]: Since `FunctionalQuadraticPerturb` inherits from `Functional` the `super` call is going to execute the code in `Functional.__init__`. So if the test there is sufficient (which I think it is), then this should be fine.

File: odl/solvers/functional/functional.py [hunk 3]
```diff
@@ -1151,10 +1164,10 @@ def __init__(self, left, right):
         if not isinstance(right, Functional):
             raise TypeError('`right` {} is not a `Functional` instance'
                             ''.format(right))
-
+        #HELP: Check ranges?
```
Comment (aringh) [inline odl/solvers/functional/functional.py]: Should not be needed; it is done in `OperatorPointwiseProduct.__init__`.

File: odl/solvers/functional/functional.py (hunk 2 above)
Comment (miallo) [reply odl/solvers/functional/functional.py]: It has been since forever that I worked with subclasses... 😅 (Which is probably a good indicator for my terrible style of programming)

File: odl/solvers/functional/functional.py [hunk 4]
```diff
@@ -1216,9 +1229,11 @@ def __init__(self, dividend, divisor):
 
         self.__dividend = dividend
         self.__divisor = divisor
-
+        
+        #HELP: Check ranges?
```
Comment (aringh) [inline odl/solvers/functional/functional.py]: Yes, here you would have to check the range.

File: odl/solvers/functional/functional.py [hunk 5]
```diff
@@ -1303,9 +1318,9 @@ def __init__(self, func):
         if not isinstance(func, Functional):
             raise TypeError('`func` {} is not a `Functional` instance'
                             ''.format(func))
-
+        #HELP: The range should be equal?!
```
Comment (aringh) [inline odl/solvers/functional/functional.py]: Actually, I think that here is where it becomes tricky.... if the range is complex then how should one define the convex conjugate? This is related to #590. So my current best answer is: I am not sure of what the range of this functional should be to make most sense.

File: odl/solvers/functional/functional.py (hunk 4 above)
Comment (aringh) [reply odl/solvers/functional/functional.py]: Actually, let me be a bit more specific. What needs to be checked is that the range of the two functionals are the same. Since this only inherits from `Functional`, there is no check elsewhere that will make sure that this is the case. In the same way, there is already a check that the domains are the same.

File: odl/solvers/functional/functional.py (hunk 5 above)
Comment (miallo) [reply odl/solvers/functional/functional.py]: Now we could just check for `functional.range == RealNumbers` and otherwise throw an error

File: odl/solvers/functional/functional.py [hunk 6]
```diff
@@ -64,7 +64,12 @@ def __init__(self, space, linear=False, grad_lipschitz=np.nan):
         # Cannot use `super(Functional, self)` here since that breaks
         # subclasses with multiple inheritance (at least those where both
         # parents implement `__init__`, e.g., in `ScalingFunctional`)
-        Operator.__init__(self, domain=space, range=space.field, linear=linear)
+#        print(range)
+        if range is None:
+            range = domain.field
+        elif range != domain.field:
+            linear = False
```
Comment (kohr-h) [inline odl/solvers/functional/functional.py]: After some very recent discussions (see #1331), we decided for practical reasons to go with linearity in the sense "C = R^2". That would mean here that this `elif` case should go away.

File: odl/solvers/functional/default_functionals.py [hunk 7]
```diff
@@ -453,18 +453,21 @@ class IndicatorLpUnitBall(Functional):
...
             Domain of the functional.
         exponent : int or infinity
             Specifies wich norm to use.
         """
-        super(IndicatorLpUnitBall, self).__init__(space=space, linear=False)
-        self.__norm = LpNorm(space, exponent)
+        super(IndicatorLpUnitBall, self).__init__(domain=domain, linear=False,
+                                                  range=range)
+        # HELP: The LpNorm should always have a range in the real numbers so
+        #       that it is comparable to 1, right?
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Yeah, so you can override the `Functional` default before the `super` call:
```python
if range is None:
    range = RealNumbers()
```
And then see what happens :laughing:

File: odl/solvers/functional/functional.py (hunk 5 above)
Comment (kohr-h) [reply odl/solvers/functional/functional.py]: Yes so we have that issue already now and just decide not to do anything about it. So IMO let's not raise any exceptions yet, instead do the right thing for `RealNumbers` and otherwise do whatever we were doing all along. At some point this whole complex functionals thing needs a thorough analysis and solution (or maybe not?), but this is not the time.

File: odl/solvers/functional/default_functionals.py (hunk 7 above)
Comment (miallo) [reply odl/solvers/functional/default_functionals.py]: Isn't this more like what we want to have: `self.__norm = LpNorm(domain, exponent, range=RealNumbers())`
Then the range of the class itself still can be complex if needed

Comment (kohr-h) [reply odl/solvers/functional/default_functionals.py]: Oh I thought you were in `LpNorm`, didn't realize it was the indicator. There, yes, I agree with your suggestion.

File: odl/solvers/functional/default_functionals.py [hunk 8]
```diff
@@ -59,18 +59,18 @@ class LpNorm(Functional):
...
         ----------
-        space : `DiscreteLp` or `TensorSpace`
+        domain : `DiscreteLp` or `TensorSpace`
             Domain of the functional.
         exponent : float
             Exponent for the norm (``p``).
         """
         super(LpNorm, self).__init__(
-            space=space, linear=False, grad_lipschitz=np.nan)
+            domain=domain, linear=False, grad_lipschitz=np.nan, range=range)
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: So this is step 1: allow to manually set the range. I'm wondering about step 2: the sensible default for norms and such. What about
```python
if range is None and getattr(domain, 'is_complex', False):
    range = RealNumbers()
super(...)
```

File: odl/solvers/functional/default_functionals.py [hunk 9]
```diff
@@ -500,6 +503,7 @@ def convex_conj(self):
         [BC2011] Bauschke, H H, and Combettes, P L. *Convex analysis and
         monotone operator theory in Hilbert spaces*. Springer, 2011.
         """
+        # HELP: I guess it needs `range`, too
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Yes, you'll want to use `self.range` (it makes sense IMO, and since it's a `@property` you can't pass arguments).

File: odl/solvers/functional/functional.py (hunk 1 above)
Comment (miallo) [reply odl/solvers/functional/functional.py]: You are right and if I am not mistaken, it would change in a nontrivial way. My guess for the upper bound would be something like `func.grad_lipschitz * ||constant||_\infty`? Should I implement that or will it be scrapped anyways?

File: odl/solvers/functional/default_functionals.py [hunk 10]
```diff
@@ -1768,16 +1777,20 @@ def __init__(self, operator=None, vector=None, constant=0):
...
             domain = vector.space
+            range = vector.space.field
 
         if (operator is not None and vector is not None and
                 vector not in operator.domain):
             raise ValueError('domain of `operator` and space of `vector` need '
                              'to match')
-
+        # HELP: operator.range == operator.domain to be able to evaluate .inner
+        #       => functional.range = operator.domain.field ?
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: True, this check is missing (or delayed until evaluation)

File: odl/solvers/functional/default_functionals.py [hunk 11]
```diff
@@ -2154,9 +2168,10 @@ def __init__(self, space, outer_exp=1, singular_vector_exp=2):
         >>> norm(space.one())
         inf
         """
+        # HELP: range should be RealNumbers since comparable?
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: It's a norm, so yes

File: odl/solvers/functional/default_functionals.py [hunk 12]
```diff
@@ -2262,8 +2277,10 @@ def __init__(self, functional, sigma=1.0):
         >>> l1_norm = odl.solvers.L1Norm(space)
         >>> smoothed_l1 = MoreauEnvelope(l1_norm)
         """
+        # HELP: Does this only makes sense, if 
+        #       functional.range == RealNumbers()? Otherwise raise Error?
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: No, there's a definition adapted for complex numbers, where the minimum is taken over the real part of the inner product only. Leave this as it is for now.

Comment (kohr-h) [reply odl/solvers/functional/default_functionals.py]: > Leave this as it is for now.

That is, do add the `range` parameter.

File: odl/solvers/functional/functional.py (hunk 1 above)
Comment (miallo) [reply odl/solvers/functional/functional.py]: Wait... This only works for linear operators, correct? So just ignore it and make it `NaN`?

File: odl/solvers/functional/default_functionals.py [hunk 13]
```diff
@@ -223,6 +230,9 @@ def __init__(self, vfspace, exponent=None):
             0 and 1 are currently not supported due to numerical
             instability. Infinity gives the supremum norm.
             Default: ``vfspace.exponent``, usually 2.
+        range : `Field`, optional
+            Range of the functional. The default `None` will be treated as
+            `vfspace.field`.
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: We need to decide on a default value here. Either we go for `RealNumbers` for all functionals that can only return real values anyway, or we go for `domain.field`. I'm personally not quite sure which convention is the best, but I guess the msot restrictive would be best, which indicates that we should go for `RealNumbers` whenever the functional is purely real valued.

File: odl/solvers/functional/default_functionals.py [hunk 14]
```diff
@@ -453,25 +466,36 @@ class IndicatorLpUnitBall(Functional):
...
         self.__exponent = float(exponent)
+        self.__range = range
 
     @property
     def exponent(self):
         """Exponent corresponding to the norm."""
         return self.__exponent
 
+    @property
+    def range_(self):
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: This is not needed. `IndicatorLpUnitBall` inherits from `Operator` and hence already has a `range` property, which is the one that should be used.

File: odl/solvers/functional/default_functionals.py [hunk 15]
```diff
@@ -599,15 +627,18 @@ class L2Norm(LpNorm):
...
+    def __init__(self, domain, range=None):
         """Initialize a new instance.
 
         Parameters
         ----------
-        space : `DiscreteLp` or `TensorSpace`
+        domain : `DiscreteLp` or `TensorSpace`
             Domain of the functional.
+        range : `Field`, optional
+            Range of the functional. The default `None` will be treated as
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: The standard notation is

```
Range of the functional. Default: `RealNumbers`
```

File: odl/solvers/functional/default_functionals.py [hunk 16]
```diff
@@ -1547,7 +1599,7 @@ def proximal(self):
         See Also
         --------
         odl.solvers.nonsmooth.proximal_operators.\
-proximal_convex_conj_kl_cross_entropy :
+        proximal_convex_conj_kl_cross_entropy :
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: This will break the link. It has to stay un-indented

File: odl/solvers/functional/default_functionals.py [hunk 17]
```diff
@@ -1655,7 +1707,8 @@ def __init__(self, *functionals):
         domain = ProductSpace(*domains)
         linear = all(func.is_linear for func in functionals)
 
-        super(SeparableSum, self).__init__(space=domain, linear=linear)
+        super(SeparableSum, self).__init__(domain=domain, linear=linear,
+                                           range=functionals[0].range)
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: I'm not sure about this choice. One would perhaps expect that real+complex should be complex, but with this implementation it depends on the order.

File: odl/solvers/functional/default_functionals.py [hunk 18]
```diff
@@ -2239,10 +2307,10 @@ class MoreauEnvelope(Functional):
     References
     ----------
     .. _Proximal Algorithms: \
-https://web.stanford.edu/~boyd/papers/pdf/prox_algs.pdf
+    https://web.stanford.edu/~boyd/papers/pdf/prox_algs.pdf
     """
 
-    def __init__(self, functional, sigma=1.0):
+    def __init__(self, functional, sigma=1.0, range=RealNumbers()):
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Should be None

File: odl/solvers/functional/default_functionals.py [hunk 19]
```diff
@@ -2364,6 +2441,7 @@ def __init__(self, space, gamma):
         >>> abs(huber_norm(x) - l1_norm(x)) < tol
         True
         """
+
```
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: We typically do not add a newline after the docstring like this

**Review by adler-j — CHANGES_REQUESTED** [review]
I did a quick review. Overall nice changes, but some stylistic stuff etc needs to be adressed.

The main issue with this is that it's a sweeping change but without any tests added. I'd prefer proper testing to be performed.

File: odl/solvers/functional/default_functionals.py (hunk 17 above)
Comment (miallo) [reply odl/solvers/functional/default_functionals.py]: I understand your reasoning and personally I would agree with you. I just thought I went with the convention of operators, where the domain (including dtype) has to be the same. But you are certainly right, that this should not depend on the order of the ops. I will think about it

File: odl/set/sets.py [hunk 20]
```diff
@@ -428,10 +428,7 @@ def __hash__(self):
 
     def element(self, inp=None):
         """Return a real number from ``inp`` or from scratch."""
-        if inp is not None:
-            return float(inp)
-        else:
-            return 0.0
+        return float(getattr(inp, 'real', 0.0))
```
Comment (miallo) [inline odl/set/sets.py]: Is this the way to go?

File: odl/solvers/functional/default_functionals.py [hunk 21]
```diff
@@ -1702,13 +1684,21 @@ def __init__(self, *functionals):
...
                 isinstance(functionals[1], Integral)):
             functionals = [functionals[0]] * functionals[1]
+        if any(functional.domain is ComplexNumbers()
+                for functional in functionals):
+            range = ComplexNumbers()
+        elif any(functional.domain is RealNumbers()
+                for functional in functionals):
+            range = RealNumbers()
+        else:
+            range = Integers()
```
Comment (miallo) [inline odl/solvers/functional/default_functionals.py]: Should we even consider Integer functionals? Is there any application for them?

File: odl/solvers/functional/default_functionals.py [hunk 22]
```diff
@@ -1684,11 +1684,11 @@ def __init__(self, *functionals):
         if (len(functionals) == 2 and
                 isinstance(functionals[1], Integral)):
             functionals = [functionals[0]] * functionals[1]
-        if any(functional.domain is ComplexNumbers()
+        if any(functional.range == ComplexNumbers()
```
Comment (miallo) [inline odl/solvers/functional/default_functionals.py]: Would it make sense to give all sets an order-parameter? Something like Integers().hirachy_number = 0, RealNumbers().hirachy_number = 1, ComplexNumbers().hirachy_number = 2? Then the _highest priority_ of range would be easy without all those comparisons

File: odl/solvers/functional/functional.py [hunk 23]
```diff
@@ -48,23 +49,33 @@ class Functional(Operator):
     <http://odlgroup.github.io/odl/guide/in_depth/functional_guide.html>`_.
     """
 
-    def __init__(self, space, linear=False, grad_lipschitz=np.nan):
+    def __init__(self, domain, linear=False, grad_lipschitz=np.nan,
```
Comment (miallo) [inline odl/solvers/functional/functional.py]: Should we change the order to `(self, domain, range, ...)`?
-  This might lead to incompatibilities with existing code like `Functional(space, False)`. Yes, I know that code like this is considered to be bad practise by many, but that probably doesn't stop users from doing so...
+  On the other hand it is the order of operator arguments (and more intuitive if you ask me).
-  For the `default_functionals` it can't be done since most have non-optional other arguments.

File: odl/set/sets.py (hunk 20 above)
Comment (kohr-h) [reply odl/set/sets.py]: I wouldn't do it this way. The issue is that this maps anything which hasn't got a `.real` attribute to `0.0`. Core types like `int` or `bool` will play nicely with this, but anything else that could be mapped to a float wouldn't. For instance strings or custom classes that implement `__float__`, but not `real`.

So I'd prefer this:
```python
if inp is None:
    return 0.0
else:
    return float(getattr(inp, 'real', inp))
```

File: odl/solvers/functional/default_functionals.py [hunk 24]
```diff
@@ -223,6 +229,8 @@ def __init__(self, vfspace, exponent=None):
             0 and 1 are currently not supported due to numerical
             instability. Infinity gives the supremum norm.
             Default: ``vfspace.exponent``, usually 2.
+        range : `Field`, optional
+            Range of the functional. Default: `vfspace.field`.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: For this kind of thing you need to use double backticks. 
With single backticks, Sphinx will try to generate a hyperlink, which fails because Sphinx doesn't know the type of `vfspace`. Single backticks are usable for class names and glossary terms mostly, other things that are not referable need to go in double backticks.

File: odl/solvers/functional/default_functionals.py [hunk 25]
```diff
@@ -324,8 +332,8 @@ def convex_conj(self):
 
     def __repr__(self):
         """Return ``repr(self)``."""
-        return '{}({!r}, exponent={})'.format(self.__class__.__name__,
-                                              self.domain,
+        return '{}({!r} -> {!r}, exponent={})'.format(self.__class__.__name__,
+                                              self.domain, self.range,
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: We try, whenever possible, to make `repr` return a string that is a valid Python expression. So please add `range` as a second argument.
Note that there's a quite huge overhaul of all this repr stuff going on, see #1276 for reference. You don't need to spend too much effort on this part.

File: odl/operator/operator.py [hunk 26]
```diff
@@ -1106,8 +1105,14 @@ def __init__(self, left, right, tmp_ran=None, tmp_dom=None):
         rn(3).element([ 2.,  4.,  6.])
         """
         if left.range != right.range:
-            raise OpTypeError('operator ranges {!r} and {!r} do not match'
-                              ''.format(left.range, right.range))
+            if isinstance(left.range, Field) and \
```
Comment (kohr-h) [inline odl/operator/operator.py]: Please no backslash line continuation. To split across lines, use parentheses:
```python
if (isinstance(left.range, Field) and
        isinstance(right.range, Field)):
    # do stuff
```

File: odl/set/sets.py [hunk 27]
```diff
@@ -310,6 +310,25 @@ def field(self):
         """
         return self
 
+    def contains_set(self, other):
+        raise NotImplementedError('field {!r} does not have method '
+                                  '`contains_set`'.format(self))
```
Comment (kohr-h) [inline odl/set/sets.py]: This error would already be raised by the parent, wouldn't it?

File: odl/set/sets.py [hunk 28]
```diff
@@ -310,6 +310,25 @@ def field(self):
...
+
+    def __add__(self, other):
+        if self.contains_set(other):
+            return self
+        elif other.contains_set(self):
+            return other
+        else:
+            raise ValueError('Fields {!r} and {!r} do not include '
+                             'each other'.format(self, other))
```
Comment (kohr-h) [inline odl/set/sets.py]: I'm skeptical towards this. It doesn't fulfill the rules for set addition, and just as a shortcut for something used in `OperatorSum`, it shouldn't introduce an API change for `Field`.

File: odl/solvers/functional/default_functionals.py [hunk 29]
```diff
@@ -59,18 +60,23 @@ class LpNorm(Functional):
...
-        space : `DiscreteLp` or `TensorSpace`
+        domain : `DiscreteLp` or `TensorSpace`
             Domain of the functional.
         exponent : float
             Exponent for the norm (``p``).
+        range : `Field`, optional
+            Range of the functional. Default: `RealNumbers`.
         """
+        if range is None:
+            range = RealNumbers()
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: How about 
```python
def __init__(self, domain, exponent, range=RealNumbers()):
    ...
```
? I feel that the current variant is unnecessarily indirect. `None` is usually a proxy for `some_value_that_depends_on_other_stuff_or_must_be_computed`

File: odl/solvers/functional/default_functionals.py [hunk 30]
```diff
@@ -2129,18 +2181,19 @@ class IndicatorNuclearNormUnitBall(Functional):
...
-        space : `ProductSpace` of `ProductSpace` of `TensorSpace`
+        domain : `ProductSpace` of `ProductSpace` of `TensorSpace`
             Domain of the functional.
         outer_exp : {1, 2, inf}, optional
             Exponent for the outer norm.
         singular_vector_exp : {1, 2, inf}, optional
             Exponent for the norm for the singular vectors.
-
+        range : `Field`, optional
+            Range of the functional. Default: `domain.field`.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Blank line needed here

File: odl/solvers/functional/functional.py (hunk 23 above)
Comment (kohr-h) [reply odl/solvers/functional/functional.py]: > Should we change the order to (self, domain, range, ...)?

IMO definitely. Mostly because `Operator` works the same way, and I agree that it's the natural order of arguments.



> This might lead to incompatibilities with existing code like Functional(space, False). Yes, I know that code like this is considered to be bad practise by many, but that probably doesn't stop users from doing so...

Well I haven't seen any direct subclassing of `Functional` out there in the wild, so I wouldn't be too concerned with breaking shitty code like this.



> For the default_functionals it can't be done since most have non-optional other arguments.

True, but I feel this should be the special case, not the rule.

File: odl/solvers/functional/functional.py [hunk 31]
```diff
@@ -686,7 +698,12 @@ def __init__(self, func, vector):
                             ''.format(func))
 
         OperatorRightVectorMult.__init__(self, operator=func, vector=vector)
-        Functional.__init__(self, space=func.domain)
+        # TODO: Implement grad_lipschitz? If func.is_linear:
+        #           grad_lipschitz = func.grad_lipschitz * ||constant||_\infty
+        #       else: grad_lipschitz = np.nan
```
Comment (kohr-h) [inline odl/solvers/functional/functional.py]: Nah, I wouldn't bother. The odds are that `grad_lipschitz` will be removed.

File: odl/solvers/functional/functional.py [hunk 32]
```diff
@@ -929,8 +948,10 @@ def __init__(self, left, right):
             raise TypeError('`func` {} is not a `Functional` instance'
                             ''.format(right))
 
+        # HELP: What should happen if left.range != right.range?
```
Comment (kohr-h) [inline odl/solvers/functional/functional.py]: IMO either nothing, or raise early whatever error adding the convex conjugates would raise. I tend towards not doing anything and letting `convex_conj` fail, at least if the error message is good enough to tell users what exactly went wrong.

Another way to play this would be to strictly require `RealNumbers` as `range` for both, since from the mathematical definition it wouldn't make sense otherwise. However, as in the convex conjugate there's probably an adapted definition for complex-valued functionals, so we probably shouldn't close the door here.

File: odl/solvers/functional/functional.py [hunk 33]
```diff
@@ -1151,10 +1171,11 @@ def __init__(self, left, right):
         if not isinstance(right, Functional):
             raise TypeError('`right` {} is not a `Functional` instance'
                             ''.format(right))
+        range = left.range + right.range
```
Comment (kohr-h) [inline odl/solvers/functional/functional.py]: See earlier comment, let's not add such cleverness.

File: odl/solvers/functional/functional.py [hunk 34]
```diff
@@ -1211,14 +1232,14 @@ def __init__(self, dividend, divisor):
             raise TypeError('`divisor` {} is not a `Functional` instance'
                             ''.format(divisor))
 
-        if dividend.domain != divisor.domain:
-            raise ValueError('domains of the operators do not match')
+        range = dividend.range + divisor.range
```
Comment (kohr-h) [inline odl/solvers/functional/functional.py]: Same here

File: odl/ufunc_ops/ufunc_ops.py [hunk 35]
```diff
@@ -326,7 +326,7 @@ def __init__(self, field):
                              "".format(name))
 
         linear = name in LINEAR_UFUNCS
-        Functional.__init__(self, space=field, linear=linear)
+        Functional.__init__(self, domain=field, linear=linear)
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/ufunc_ops/ufunc_ops.py]: It's just for clarity. 

If you write `Functional.__init__(domain, range, linear)` it's obvious what each of the arguments does, and keywords just add noise: `Functional.__init__(domain=domain, range=range, linear=linear)`.
On the other hand, something like `Functional.__init__(field, True)` doesn't convey any information about the purpose of the arguments, so here `Functional.__init__(domain=field, linear=True)` is much clearer.

But why is this related to tests?

File: odl/solvers/functional/default_functionals.py (hunk 29 above)
Comment (adler-j) [reply odl/solvers/functional/default_functionals.py]: This only works if we want `RealNumbers` to be forever immutable. Are we sure about that? (Id be fine with it)

File: odl/solvers/functional/default_functionals.py (hunk 25 above)
Comment (kohr-h) [reply odl/solvers/functional/default_functionals.py]: Actually, after looking at the failing doctest (output `L2Norm(rn(3) -> RealNumbers())` doesn't match expected output `L2Norm(rn(3))`, something like this), I'd suggest not changing anything in `__repr__`. That's suboptimal for the non-standard case but at least doesn't break the printout for the default case.

File: odl/set/sets.py (hunk 20 above)
Comment (miallo) [reply odl/set/sets.py]: I hadn't thought about that... That is a nice solution!

File: odl/solvers/functional/default_functionals.py (hunk 18 above)
Comment (kohr-h) [reply odl/solvers/functional/default_functionals.py]: When did we make that decision? I vaguely remember a discussion.

File: odl/solvers/functional/default_functionals.py (hunk 29 above)
Comment (kohr-h) [reply odl/solvers/functional/default_functionals.py]: Yeah, we can make that clearer by adding `__slots__` to `RealNumbers` and `ComplexNumbers`.

File: odl/ufunc_ops/ufunc_ops.py (hunk 35 above)
[the remark this answers is not in the document]
Comment (miallo) [reply odl/ufunc_ops/ufunc_ops.py]: I stumbled across it in the test results and wanted to make sure that someone else takes over when I am done with implementing most of the stuff

File: odl/set/sets.py (hunk 28 above)
Comment (miallo) [reply odl/set/sets.py]: I am adding fields, not sets and so I thought I went with the _Minkowski addition_ `A + B = {a + b | a in A, b in B}` (which at least for the implemented numbers is trivial). I actually haven't thought about other sets.

What would be your suggestion to approach this? At first I did something like `if isinstance(a.range, ComplexNumbers) or isinstance(b.range, ComplexNumbers): return ComplexNumbers(); elif ...` but that was rather unpleasant to look at...

Comment (miallo) [reply odl/set/sets.py]: I see... The spaces are homomorph but not isomorph. You were right all along...

Comment (kohr-h) [reply odl/set/sets.py]: :rofl: You went on quite a journey there, my mention of finite fields was more like an example.. In some sense you proved my minimal point that it's not trivial to implement set addition and keep it in line with mathematics for arbitrary fields.

But it doesn't seem to be necessary here anyway.

File: odl/set/sets.py [hunk 36]
```diff
@@ -428,10 +428,10 @@ def __hash__(self):
 
     def element(self, inp=None):
         """Return a real number from ``inp`` or from scratch."""
-        if inp is not None:
-            return float(inp)
-        else:
+        if inp is None:
             return 0.0
+        else:
+            return float(getattr(inp, 'real', inp))
```
Comment (kohr-h) [inline odl/set/sets.py]: Besides the implementation, I'm wondering if we really want this implicit conversion. I'd actually prefer to let this fail for objects that don't implement `__float__` but do implement `real`.
There's also a weird side effect for objects that implement both `__float__` and `real`, namely that `real` is preferred over `__float__`. That doesn't seem quite right to me.
Other opinions, @aringh, @adler-j?

Comment (aringh) [reply odl/set/sets.py]: In this last case, should `ImagPart(RealNumbers().element(1))` give `0` or should it fail? The latter represents seeing `RealNumbers` and `ComplexNumbers` as two different entities, while the former somehow implicitly assumes that the real numbers are seen as embedded in the complex numbers, just with imaginary part 0. Which one is preferred?

Comment (kohr-h) [reply odl/set/sets.py]: I'd take Python (and NumPy) as a guide here:
```python
>>> x = 1.0
>>> x.imag
0.0
>>> x = np.array(1.0)
>>> x.imag
array(0.)
>>> float(1 + 1j)
...
TypeError: can't convert complex to float
>>> np.array(1 + 1j, dtype=float)
...
TypeError: can't convert complex to float
```

Comment (adler-j) [reply odl/set/sets.py]: Some quick trials gives:

```
>>> (1).imag
0
```

Hence, `ImagPart(RealNumbers().element(1))` should return 0.

File: odl/solvers/functional/default_functionals.py [hunk 37]
```diff
@@ -1647,15 +1678,23 @@ def __init__(self, *functionals):
         >>> f_sum = odl.solvers.SeparableSum(l1, 5)
         """
         # Make a power space if the second argument is an integer
-        if (len(functionals) == 2 and
-                isinstance(functionals[1], Integral)):
+        if (len(functionals) == 2 and isinstance(functionals[1], Integral)):
             functionals = [functionals[0]] * functionals[1]
+        for func1 in functionals:
```
Comment (miallo) [inline odl/solvers/functional/default_functionals.py]: I guess
```Python
from functools import reduce
...
range = reduce((lambda x, func: x if func.range.contains_set(x) else func.range), functionals)
```
would be a more elegant way, but needs the import. Any comments?

Comment (miallo) [reply odl/solvers/functional/default_functionals.py]: Now that I think about it, I guess this is the way to do it (because it doesn't need the import):
```Python
range = RealNumbers()
for func in functionals:
    range = func.range if func.range.contains(range) else range
```
(Or maybe set `range = None` in the beginning?)

Comment (miallo) [reply odl/solvers/functional/default_functionals.py]: Or maybe it is reasonable to implement a method for the unification of sets? Something like
```Python
class Set {...
    def unified_with(self, other):
        """This is the default implementation for the unification of two sets
        It only supports equal sets
        """
        if type(self) == type(other):
            return other
        else:
            raise NotImplementedError('Cannot get the unified field for {} and'
                                      '{}'.format(self, other))
```
And the Number implementation
```Python
    def unified_with(self, other):
        """Returns the unified Field of `self` and `other`"""
        return other if other.contains_set(self) else self
```
That way new fields could easily implement their own method.
If we would end up doing so you might want to discuss the name...

Comment (kohr-h) [reply odl/solvers/functional/default_functionals.py]: I'd suggest the simplest possible solution:
```python
range = RealNumbers()
for func in functionals:
    if func.range.contains(range):
        range = func.range
```

File: odl/operator/default_ops.py [hunk 38]
```diff
@@ -182,7 +182,7 @@ class LinCombOperator(Operator):
         LinCombOperator(a, b)([x, y]) == a * x + b * y
     """
 
-    def __init__(self, space, a, b):
+    def __init__(self, space, a, b, rangeType=None):
```
Comment (miallo) [inline odl/operator/default_ops.py]: Oh, that wasn't meant to be pushed... I used it for a mapping-oparator from R2 to C...

**Comment by kohr-h** [comment]
> Merged from odl main branch

Two things:

1. We don't do merge from master, please rebase instead. That keeps a nice and clean history. (If you need to know how it works, just ask.)
2. We have [some guidelines](https://odlgroup.github.io/odl/dev/gitwash/development_workflow.html#the-commit-message) regarding commit messages that we try to follow ourselves (though sometimes we get sloppy, too :wink:). It makes the list of commits much more digestible if you see at a glance which commit does what.
   Short version:
   - Prefix with `BUG:`, `MAINT:`, `STY:`, `TST:` etc.
   - Use imperative style ("Fix bug", not "Fixed bug")
   - Fit first line in ~50 characters, then go wild after a blank line

I guess this PR is gonna end up as one big squashed commit since it's been going a bit back and forth (no issue with that), so we'll likely fix it then.

**Comment by adler-j** [comment]
So what's the status of this? I'd love for this to get finalized!

**Comment by miallo** [comment]
> So what's the status of this?

As I mentioned somewhere before: I have never written any tests in python and I would rather have someone else do them since I am not 100% confident with what is to be expected for the default values (expecially for things like `IndicatorLpUnitBall.convex_conj`). Otherwise I hope I found most of the 'obvious' bugs for now...

**Comment by adler-j** [comment]
I'm currently very busy leading up to the vacation, but I can do that after summer. Thanks again for the contribution!
